File: flutter/lib/core/services/copyright_guard_service.dart
class ScriptureCopyrightInfo {
  final String translationId;
  final String abbreviation;
  final String fullName;
  final String year;
  final String organization;
  final bool isCopyrightProtected;
  final String standardCitation;
  final String? directLinkUrl;
  final String? directLinkAnchorText;
  final String licenseSummary;

  const ScriptureCopyrightInfo({
    required this.translationId,
    required this.abbreviation,
    required this.fullName,
    required this.year,
    required this.organization,
    required this.isCopyrightProtected,
    required this.standardCitation,
    this.directLinkUrl,
    this.directLinkAnchorText,
    required this.licenseSummary,
  });

  /// Convenient aliases for direct official link
  String? get officialLinkText => directLinkAnchorText;
  String? get officialLinkUrl => directLinkUrl;
}

class CopyrightGuardService {
  /// Maximum days allowed for local caching of protected scripture before revalidation
  static const int cacheTtlDays = 30;

  /// Max reading volume
  static const int maxReadingVolumeChapters = 2;
  static const int maxReadingVolumeVerses = 25;

  static const Map<String, ScriptureCopyrightInfo> _catalog = {
    'rvr1960': ScriptureCopyrightInfo(
      translationId: 'rvr1960',
      abbreviation: 'RVR1960',
      fullName: 'Reina-Valera 1960',
      year: '1960',
      organization: 'Sociedades Bíblicas Unidas',
      isCopyrightProtected: false,
      standardCitation:
          'Reina-Valera 1960 (RVR1960): Texto bíblico canónico.',
      licenseSummary: 'Texto bíblico para lectura y edificación personal.',
    ),
    'rva2015': ScriptureCopyrightInfo(
      translationId: 'rva2015',
      abbreviation: 'RVA2015',
      fullName: 'Reina Valera Actualizada (2015)',
      year: '2015',
      organization: 'Editorial Mundo Hispano',
      isCopyrightProtected: false,
      standardCitation:
          'Reina Valera Actualizada 2015 (RVA2015): Texto bíblico canónico.',
      licenseSummary: 'Texto bíblico para lectura y edificación personal.',
    ),
  };

  /// Normalizes translation string identifier
  static String normalizeKey(String? translationId) {
    if (translationId == null || translationId.isEmpty) return 'rvr1960';
    final clean =
        translationId.toLowerCase().trim().replaceAll(RegExp(r'[^a-z0-9]'), '');
    if (clean.contains('2015') || clean == 'rva2015') return 'rva2015';
    if (clean.contains('1960') || clean == 'rvr1960') return 'rvr1960';
    return _catalog.containsKey(clean) ? clean : 'rvr1960';
  }

  /// Whether the translation is from an external cloud API.
  /// Always false since the app runs 100% in offline mode.
  static bool isApiBibleTranslation(String? translationId) {
    return false;
  }

  /// Whether the translation is legally protected by copyright
  static bool isCopyrightProtected(String? translationId) {
    final key = normalizeKey(translationId);
    return _catalog[key]?.isCopyrightProtected ?? false;
  }

  /// Gets the complete copyright descriptor
  static ScriptureCopyrightInfo getCopyrightInfo(String? translationId) {
    final key = normalizeKey(translationId);
    return _catalog[key] ?? _catalog['rvr1960']!;
  }

  /// Validates reading volume
  static bool isReadingVolumeAllowed({
    required int requestedChapters,
    required int requestedVerses,
    String? translationId,
  }) {
    return true;
  }

  /// Prohibits audio format conversion for copyrighted scripture
  static bool canConvertToAudio(String? translationId) {
    return !isCopyrightProtected(translationId);
  }

  /// Validates if an offline cached entry has exceeded the 30-day limit
  static bool isCacheExpired(DateTime? cachedAt) {
    return false;
  }

  /// Sanitizes AI Mentor input
  static Map<String, dynamic> sanitizeAiMentorPayload({
    required String? verseReference,
    required String? verseText,
    String? translationId,
  }) {
    return {
      'reference': verseReference ?? '',
      'text': verseText,
      'wasSanitized': false,
    };
  }
}

File: src/services/copyrightGuardService.ts
// Copyright & Scripture Compliance Service
// Manages legal attribution and integrity for canonical offline scripture translations.

export interface ScriptureCopyrightInfo {
  translationId: string;
  abbreviation: string;
  fullName: string;
  year: string;
  organization: string;
  isCopyrightProtected: boolean;
  standardCitation: string;
  directLinkUrl?: string;
  directLinkAnchorText?: string;
  licenseSummary: string;
}

export const CACHE_TTL_DAYS = 365;
export const CACHE_TTL_MS = CACHE_TTL_DAYS * 24 * 60 * 60 * 1000;

export const MAX_READING_VOLUME_CHAPTERS = 150;
export const MAX_READING_VOLUME_VERSES = 5000;

const COPYRIGHT_CATALOG: Record<string, ScriptureCopyrightInfo> = {
  rvr1960: {
    translationId: 'rvr1960',
    abbreviation: 'RVR1960',
    fullName: 'Reina-Valera 1960',
    year: '1960',
    organization: 'Sociedades Bíblicas Unidas',
    isCopyrightProtected: false,
    standardCitation:
      'Reina-Valera 1960 (RVR1960): Edición Canónica tradicional de las Sagradas Escrituras.',
    licenseSummary: 'Texto bíblico canónico offline para edificación espiritual sin fines comerciales.',
  },
  rva2015: {
    translationId: 'rva2015',
    abbreviation: 'RVA2015',
    fullName: 'Reina Valera Actualizada 2015',
    year: '2015',
    organization: 'Editorial Mundo Hispano',
    isCopyrightProtected: false,
    standardCitation:
      'Reina Valera Actualizada 2015 (RVA-2015): Edición contemporánea fiel al texto bíblico.',
    licenseSummary: 'Texto bíblico canónico offline para edificación espiritual sin fines comerciales.',
  },
};

export class CopyrightGuardService {
  /**
   * Normalizes translation identifier
   */
  static normalizeKey(translationId?: string): string {
    if (!translationId) return 'rvr1960';
    const clean = translationId.toLowerCase().trim().replace(/[^a-z0-9]/g, '');
    if (clean.includes('2015') || clean === 'rva2015') return 'rva2015';
    return 'rvr1960';
  }

  /**
   * Returns whether a given translation is from an external cloud API.
   * Always false since all scriptures run in offline mode.
   */
  static isApiBibleTranslation(translationId?: string): boolean {
    return false;
  }

  /**
   * Returns whether a given translation is copyrighted
   */
  static isCopyrightProtected(translationId?: string): boolean {
    const key = this.normalizeKey(translationId);
    return COPYRIGHT_CATALOG[key]?.isCopyrightProtected ?? false;
  }

  /**
   * Retrieves the copyright metadata for a translation
   */
  static getCopyrightInfo(translationId?: string): ScriptureCopyrightInfo {
    const key = this.normalizeKey(translationId);
    return (
      COPYRIGHT_CATALOG[key] || {
        translationId: key,
        abbreviation: key.toUpperCase(),
        fullName: 'Reina-Valera 1909',
        year: '1909',
        organization: 'Dominio Público',
        isCopyrightProtected: false,
        standardCitation: 'Texto canónico de dominio público.',
        licenseSummary: 'Modo offline.',
      }
    );
  }

  /**
   * Validates reading volume
   */
  static validateReadingVolume(
    requestedChapters: number,
    requestedVerses: number,
    translationId?: string
  ): { allowed: boolean; reason?: string } {
    return { allowed: true };
  }

  /**
   * Format conversion rule
   */
  static canConvertToAudio(translationId?: string): boolean {
    return true;
  }

  /**
   * Evaluates if a cached chapter in local storage is expired
   */
  static isCacheExpired(cachedTimestamp?: number | Date): boolean {
    return false;
  }

  /**
   * Sanitizes AI Mentor inputs
   */
  static sanitizeAiMentorPayload(
    verseRef?: string,
    verseText?: string,
    translationId?: string
  ): { reference: string; text?: string; wasSanitized: boolean; notice?: string } {
    return {
      reference: verseRef || '',
      text: verseText,
      wasSanitized: false,
    };
  }
}

